lib/ruby_smb/dcerpc.rb

```diff
@@ -55,35 +56,267 @@ module Dcerpc
+          auth = type1_message.serialize
+        when RPC_C_AUTHN_GSS_KERBEROS, RPC_C_AUTHN_NETLOGON, RPC_C_AUTHN_GSS_NEGOTIATE
+        when RPC_C_AUTHN_GSS_KERBEROS, RPC_C_AUTHN_NETLOGON, RPC_C_AUTHN_GSS_NEGOTIATE, RPC_C_AUTHN_GSS_SCHANNEL
+          # TODO
+          raise NotImplementedError
+        else
+          raise ArgumentError, "Unsupported Auth Type: #{options[:auth_type]}"
+        end
+        add_auth_verifier(bind_req, auth, options[:auth_type], options[:auth_level])
       end
-      unless dcerpc_response.pdu_header.ptype == PTypes::BIND_ACK
-        raise Error::BindError, "Not a BindAck packet"
+
+      send_packet(bind_req)
+      begin
+        dcerpc_response = recv_struct(BindAck)
+      rescue Error::InvalidPacket
+        raise Error::BindError # raise the more context-specific BindError
       end
+      # TODO: see if BindNack response should be handled too
 
       res_list = dcerpc_response.p_result_list
       if res_list.n_results == 0 ||
          res_list.p_results[0].result != BindAck::ACCEPTANCE
         raise Error::BindError,
           "Bind Failed (Result: #{res_list.p_results[0].result}, Reason: #{res_list.p_results[0].reason})"
       end
-      @tree.client.max_buffer_size = dcerpc_response.max_xmit_frag
+      self.max_buffer_size = dcerpc_response.max_xmit_frag
+      @call_id = dcerpc_response.pdu_header.call_id
+
+      if options[:auth_level] && options[:auth_level] != RPC_C_AUTHN_LEVEL_NONE
+        # The number of legs needed to build the security context is defined
+        # by the security provider
+        # (see [2.2.1.1.7 Security Providers](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rpce/d4097450-c62f-484b-872f-ddf59a7a0d36))
+        case options[:auth_type]
+        when RPC_C_AUTHN_WINNT
+          send_auth3(dcerpc_response, options[:auth_type], options[:auth_level])
+        when RPC_C_AUTHN_GSS_KERBEROS, RPC_C_AUTHN_NETLOGON, RPC_C_AUTHN_GSS_NEGOTIATE
+          # TODO
+          raise NotImplementedError
+        end
+      end
+
       dcerpc_response
     end
 
+    def max_buffer_size=(value)
+      @tree.client.max_buffer_size = value
+    end
+
+    # Receive a packet from the remote host and parse it according to `struct`
+    #
+    # @param struct [Class] the structure class to parse the response with
+    def recv_struct(struct)
+      raw_response = read
+      begin
+        response = struct.read(raw_response)
+      rescue IOError
+        raise Error::InvalidPacket, "Error reading the #{struct} response"
+      end
+      unless response.pdu_header.ptype == struct::PTYPE
+        raise Error::InvalidPacket, "Not a #{struct} packet"
+      end
+
+      response
+    end
+
+    # Send a packet to the remote host
+    #
+    # @param packet [BinData::Record] the packet to send
+    # @raise [Error::CommunicationError] if socket-related error occurs
+    def send_packet(packet)
+      write(data: packet.to_binary_s)
+      nil
+    end
```

Comment:
These three functions were added for API compatibility so the other methods could be consolidated here.